tools/install-source/InstallSourcesLib/XamarinSourcesPathMangler.cs

```diff
@@ -0,0 +1,185 @@
+			}
+			return src;
+		}
+
+		/// <summary>
+		/// Returns the source path for the bindings that were manually done.
+		/// </summary>
+		/// <returns>The source path for manual source.</returns>
+		/// <param name="path">Path.</param>
+		string GetSourcePathForManualSource(string path)
+		{
+			var removalPath = $"/{FrameworkPath}/";
+			var srcFolder = Path.Combine ("src", FrameworkPath.Remove(FrameworkPath.IndexOf (".framework", StringComparison.InvariantCulture))) + Path.DirectorySeparatorChar;
+			var pos = path.IndexOf (removalPath, StringComparison.InvariantCulture);
+			var src = path.Remove (0, pos + FrameworkPath.Length + 2);
+			pos = src.IndexOf (srcFolder, StringComparison.InvariantCulture);
+			if (pos >= 0)
+				src = src.Remove (0, src.IndexOf (srcFolder, StringComparison.InvariantCulture) + srcFolder.Length);
+			else // we are dealing with a correct path (this does happen on the mac side)
+				src = src.Remove (0, src.IndexOf ("/src/", StringComparison.InvariantCulture) + "/src/".Length);
+			src = Path.Combine (XamarinSourcePath, src);
+			return src;
+		}
+
+		public string GetSourcePath(string path)
+		{
+			if (IsRunrime (path))
+				return null; // TODO: Do we add this guys?
+			// decide what type of path we are dealing with and return the correct source.
+			if (IsGeneratedPath (path)) {
+				return GetSourcePathForGeneratedPath (path);
+			}
+			if (IsManualSource (path))
+				return GetSourcePathForManualSource (path);
+			return GetSourcePathForNativeType (path);
+		}
+
+		string GetTargetPathForGeneratedPath (string path)
+		{
+			var subpath = (InstallDir.Contains ("Xamarin.iOS.framework")) ? "native" : "full";
+			var pos = path.IndexOf (subpath, StringComparison.InvariantCulture);
+			if (pos >= 0) {
+				var relativePath = path.Remove (0, pos + subpath.Length + 1);
+				return Path.Combine (InstallDir, "src", FrameworkPath.Remove (FrameworkPath.IndexOf (".framework", StringComparison.InvariantCulture)), relativePath);
+			}
+			return null;
+		}
+
+		string GetTargetPathForManualSource (string path)
+		{
+			var pos = path.IndexOf (XamarinSourcePath, StringComparison.InvariantCulture);
+			if (pos >= 0) {
+				var relativePath = path.Remove (0, pos + XamarinSourcePath.Length);
+				return Path.Combine (InstallDir, "src", FrameworkPath.Remove (FrameworkPath.IndexOf (".framework", StringComparison.InvariantCulture)), relativePath);
+			}
+			return null;
+		}
+
+		string GetTargetPathForNativeType (string path)
+		{
+			var pos = path.IndexOf (NativeTypeSubpath, StringComparison.InvariantCulture);
+			if (pos >= 0) {
+				var relativePath = path.Remove (0, pos);
+				return Path.Combine (InstallDir, "src", FrameworkPath.Remove (FrameworkPath.IndexOf (".framework", StringComparison.InvariantCulture)), relativePath);	
+			}
+			return null;
+		}
+
+		public string GetTargetPath (string path)
+		{
+			if (IsRunrime (path))
+				return null; // TODO: Do we add this guys?
```

Comment:
Please consider converting TODOs into bugzilla bugs if they still open when you land. TODO in code get forgotten.